0001-dinerojs-v2.md

```diff
@@ -0,0 +1,188 @@
+
+`convert` should no longer expect a REST API and encapsulate the fetching behavior. Instead, it should expect a Promise which resolves to a `Rates` object.
+
+```js
+const rates = new Promise(resolve =>
+  resolve({
+    EUR: 0.81162
+  })
+);
+
+Dinero({ amount: 500 }).convert("EUR", { rates });
+```
+
+##### Integers & BigInts
+
+Dinero.js v1 uses integers for `amount` (though the `number` primitive), and this shouldn't change, but Dinero.js v2 should also allow for `bigint` [as soon as the proposal goes to stage 4](https://tc39.es/proposal-bigint/).
+
+In the meantime, and to support environments with no `bigint` support, Dinero.js v2 should let users provide their own `Calculator`. This way, they can use whatever library they want and map the calculator functions. This behavior should be made available through a Dinero factory, so it doesn't "pollute" the API.
+
+```js
+import { createDinero } from "@dinero.js/helpers";
+import Big from "some-bigint-library";
+
+const Dinero = createDinero({
+  calculator: {
+    add: Big.add,
+    subtract: Big.subtract
+    // ...
+  }
+});
+```
+
+A `Calculator` type should be provided to ease the creation with TypeScript:
+
+```ts
+const BigIntCalculator: Calculator = {
+  add: Big.add,
+  subtract: Big.subtract
+  // ...
+};
+```
+
+#### Currency service
+
+Currently, users must specify the `precision` of their `Dinero` objects, which is set to `2` by default. When a user creates objects with a currency which exponent is different from `2` (e.g., Iraqi dinars, Japanese yens, etc.), they must specify the `precision` by hand. One way to alleviate it is by setting a `defaultPrecision`, but this only works when one is using a single currency throughout their project.
+
+The currency service lets the user pass a `Currency` object instead of a plain string to the `currency` property when creating `Dinero` objects. `Currency` objects embark the currency's exponent, allowing the `Dinero` object to set the `precision` property automatically.
+
+```js
+import Dinero from "dinero.js";
+import { IQD } from "@dinero.js/currencies";
+
+// `precision` is set to `2` (default)
+const d1 = Dinero({ amount: 1000, currency: "IQD" });
+
+// `precision` is automatically set to `3`
+const d2 = Dinero({ amount: 1000, currency: IQD });
+```
+
+`Currency` objects could also embark the currency's symbol, allowing the library to fully own the currency formatting within the [`toFormat`](https://sarahdayan.github.io/dinero.js/module-Dinero.html#~toFormat) method and liberate itself from its dependency on the I18n API.
+
+```js
+import Dinero from "dinero.js";
+import { EUR } from "@dinero.js/currencies";
+
+const d1 = Dinero({ amount: 100, currency: "EUR" });
+const d2 = Dinero({ amount: 100, currency: EUR });
+
+d1.toFormat("$0,0.00"); // EUR100.00
```

Comment:
is there a way to still rely on the Intl result here, or should that be done with a custom currency?

Reply:
`toFormat` would be fully decoupled from I18n, there would be no wrapping. To use I18n, you'd do:

```js
d1
  .toRoundedUnit(2)
  .toLocaleString(d1.getLocale(), options)
```